File: 백준/Silver/1244. 스위치 켜고 끄기/스위치 켜고 끄기.py
n = int(input()) # 스위치의 개수
status = list(map(int,input().split())) # 스위치의 상태
students_num = int(input()) # 학생 수

students = [tuple(map(int,input().split())) for _ in range(students_num)] # 성별, 학생이 받은 스위치 수 (남학생 1 여학생 2)


for gender, num in students :
    if gender == 1 : # s남
        for i in range(num - 1, len(status), num) :
            status[i] = 1 - status[i]
    else : # 여학생
        idx = num - 1
        status[idx] = 1 - status[idx]
        
        d = 1
        
        while idx - d >= 0 and idx + d < len(status) :
            if status[idx - d] == status[idx + d] :
                status[idx - d] = 1 - status[idx - d]
                status[idx + d] = 1 - status[idx + d]
                d += 1
            else :
                break

for i in range(len(status)) :
    print(status[i], end=' ')
    
    if (i + 1) % 20 == 0 :
        print()

File: 백준/Silver/1244. 스위치 켜고 끄기/스위치 켜고 끄기.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int[] arr = new int[n + 1];
        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int students = Integer.parseInt(br.readLine());

        for (int i = 0; i < students; i++) {
            st = new StringTokenizer(br.readLine());

            int gender = Integer.parseInt(st.nextToken());
            int idx = Integer.parseInt(st.nextToken());

            // 만약 남자면
            if (gender == 1) {
                for (int j = 1; j <= n; j++) {
                    if (j % idx == 0) {
                        if (arr[j] == 1)
                            arr[j] = 0;
                        else
                            arr[j] = 1;
                    }
                }
            } else { // 여자이면
                // 1. 자기 자신은 무조건 바꿈
                arr[idx] = 1 - arr[idx];

                // 2. 대칭 검사 (범위 체크 필수!)
                for (int j = 1; idx - j >= 1 && idx + j <= n; j++) {
                    if (arr[idx - j] == arr[idx + j]) {
                        arr[idx - j] = 1 - arr[idx - j];
                        arr[idx + j] = 1 - arr[idx + j];
                    } else {
                        // 대칭이 깨지면 즉시 중단
                        break;
                    }
                }
            }

        }

        for (int i = 1; i <= n; i++) {
            System.out.print(arr[i] + " ");
            if (i % 20 == 0) { // 20번째마다 줄바꿈
                System.out.println();
            }
        }

    }

}
